rules/integrations/o365/collection_onedrive_excessive_file_downloads.toml

```diff
@@ -0,0 +1,128 @@
+- Look for multiple authentication attempts from different devices or locations within a short timeframe, which could indicate unauthorized access.
+- Investigate if other OAuth-related anomalies exist, such as consent grants for unfamiliar applications or unexpected refresh token activity.
+
+### False Positive Analysis
+
+- Verify if the user regularly downloads large batches of files as part of their job function.
+- Determine if the downloads were triggered by an authorized automated process, such as a data backup or synchronization tool.
+- Confirm if the detected OAuth application is approved for enterprise use and aligns with expected usage patterns.
+
+### Response and Remediation
+
+- If unauthorized activity is confirmed, revoke the OAuth token used and terminate active OneDrive sessions.
+- Reset the affected user's password and require reauthentication to prevent continued unauthorized access.
+- Restrict OAuth app permissions and enforce conditional access policies to limit authentication to trusted devices and applications.
+- Monitor for additional signs of compromise, such as unusual email forwarding rules, external sharing of OneDrive files, or privilege escalation attempts.
+- Educate users on OAuth phishing risks and encourage the use of **Microsoft Defender for Office 365 Safe Links** to mitigate credential-based attacks.
+- Enable continuous monitoring for OAuth authentication anomalies using **Microsoft Entra ID sign-in logs** and security tools.
+"""
+references = [
+    "https://www.volexity.com/blog/2025/02/13/multiple-russian-threat-actors-targeting-microsoft-device-code-authentication/",
+]
+risk_score = 47
+rule_id = "0e524fa6-eed3-11ef-82b4-f661ea17fbce"
+severity = "medium"
+tags = [
+    "Domain: Cloud",
+    "Domain: SaaS",
+    "Data Source: Microsoft 365",
+    "Data Source: SharePoint",
+    "Data Source: OneDrive",
+    "Use Case: Threat Detection",
+    "Tactic: Collection",
+    "Tactic: Exfiltration",
+    "Resources: Investigation Guide",
+]
+timestamp_override = "event.ingested"
+type = "esql"
+
+query = '''
+FROM logs-o365.audit-*
+| WHERE @timestamp > now() - 14 day
+| WHERE
+    event.dataset == "o365.audit" and
+
+    // filter on files downloaded from OneDrive
+    event.provider == "OneDrive" and
+    event.action == "FileDownloaded" and
+
+    // filter on OAuth authentication which encompasses device code workflow
+    o365.audit.AuthenticationType == "OAuth"
+    and event.outcome == "success"
+// bucket authentication attempts by 1 minute
+| EVAL target_time_window = DATE_TRUNC(1 minutes, @timestamp)
+| KEEP target_time_window, o365.audit.UserId, file.name, file.directory, source.ip
+
+// aggregate on unique file names and download attempts
+| STATS unique_file_count = count_distinct(file.name), download_attempt_count = count(*) BY o365.audit.UserId, file.directory, source.ip
```

Comment:
target_time_window is unused in the stats aggregation, is file.directory necessary ? 

```suggestion
| STATS unique_file_count = count_distinct(file.name), download_attempt_count = count(*) BY target_time_window, o365.audit.UserId, source.ip
```

Reply:
Removed `file.directory` as it is not necessary. Added `target_time_window` to the aggregation to ensure >= 25 files were downloaded within less than a minute for a specific user and source.